### PR TITLE
Sgx updates

### PR DESCRIPTION
**Pull request by kylebystrom**
Updates to SGX documentation and examples corresponding to the additions in Pyscf PR 945 https://github.com/pyscf/pyscf/pull/945.

### CODE REVIEW

File: examples/sgx/00-simple_sgx.py [hunk 1]
```diff
@@ -14,9 +14,37 @@
     ''',
     basis = 'ccpvdz',
 )
-mf = sgx.sgx_fit(scf.RHF(mol))
+# Direct K matrix for comparison
+mf = scf.RHF(mol)
+mf.kernel()
+
+# Using SGX for J-matrix and K-matrix
```
Comment (susilehtola) [inline examples/sgx/00-simple_sgx.py]:
```suggestion
# Using SGX for J-matrix and K-matrix, without using P-junction screening
```

File: examples/sgx/00-simple_sgx.py [hunk 2]
```diff
@@ -14,9 +14,37 @@
...
+mf = sgx.sgx_fit(scf.RHF(mol), pjs=False)
 mf.kernel()
 
 # Using RI for Coulomb matrix while K-matrix is constructed with COS-X method
 mf.with_df.dfj = True
 mf.kernel()
```
Comment (susilehtola) [inline examples/sgx/00-simple_sgx.py]: It's a bit confusing that you say COS-X here when it's SGX in the previous example. I thought COS-X has P junctions?

File: source/user/sgx.rst [hunk 3]
```diff
@@ -27,27 +27,54 @@ Any :attr:`scf.hf.SCF` object :attr:`mf` can be converted to an equivalent objec
...
+    converged SCF energy = -76.0267978618973
+    converged SCF energy = -76.0267978618973
+    converged SCF energy = -76.0267978671524
+    /home/kyle/Research/projects/devpyscf/pyscf/pyscf/sgx/sgx.py:126: UserWarning: 
+                        P-junction screening is not compatible with SGX J-matrix.
+                        Setting dfj = True. If you want to use SGX J-matrix,
+                        set pjs = False to turn off P-junction screening.
+      warnings.warn(msg)
+    converged SCF energy = -76.0267978671524
+    converged SCF energy = -76.0267685471926
```
Comment (susilehtola) [inline source/user/sgx.rst]: This looks a bit sketchy in the docs: you have 7 converged SCF energies and a five-line warning. How is one supposed to interpret this?

File: source/user/sgx.rst [hunk 4]
```diff
@@ -27,27 +27,54 @@ Any :attr:`scf.hf.SCF` object :attr:`mf` can be converted to an equivalent objec
...
+* :attr:`blockdim`: The maximum number of grid points to loop over at once. The number of grid points per batch is the minimum of :attr:`blockdim` and the maximum number of points allowed by the memory available for the calculation. The maximum memory can be adjusted by setting the :attr:`max_memory` attribute, which is initially set to :attr:`mol.max_memory`, the max memory setting for the Mole object.
+* :attr:`dfj`: If :attr:`True`, density fitting is used for the J-matrix. If :attr:`False`, SGX is used.
+* :attr:`direct_j`: If :attr:`True`, direct evaluation of the J-matrix is used (slow, for testing only).
+* :attr:`pjs`: If :attr:`True`, P-junction screening is used. Threshold determined by :attr:`direct_scf_tol`.
+
+:attr:`SGXHF` attribute:
+
+* :attr:`direct_scf_sgx`: Whether to use direct SCF within the SGX module, meaning that the J and K matrices are evaluated from the difference in the density matrix from the previous iteration.
```
Comment (susilehtola) [inline source/user/sgx.rst:62]: Incremental Fock builds typically also employ resets i.e. full Fock builds every N iterations. How is N controlled in your implementation?

If you don't do full builds, you suffer from creeping in numerical error.

File: examples/sgx/00-simple_sgx.py (hunk 2 above)
Comment (kylebystrom) [reply examples/sgx/00-simple_sgx.py]: Agreed, I fixed this.

File: source/user/sgx.rst (hunk 3 above)
Comment (kylebystrom) [reply source/user/sgx.rst]: I was just showing what the output looks like for the example script, but I guess it doesn't give any useful information and adds confusion. The UserWarning is explained below anyways. Should I just not include the output?

File: source/user/sgx.rst (hunk 4 above)
Comment (kylebystrom) [reply source/user/sgx.rst:62]: I'm not currently doing rebuilds except for when the grid level changes, and I did observe that this causes small numerical noise even for simple closed-shell organic systems (~1e-7 Eh). The direct JK evaluation in PySCF does not do rebuilds either, though I have not noticed it causing as much noise as the direct SGX version. I can add this if you want it (should be a fairly simple change), though I don't have any experience with what values of N are sensible.

Comment (susilehtola) [reply source/user/sgx.rst:62]: I think most programs do full rebuilds every 5 iterations or so.

**Comment by kylebystrom** [comment]
I added and documented a `rebuild_nsteps` attribute to control how often the JK matrix is rebuilt. I also removed the output of the example script from the SGX docs page since it didn't provide much information.

**Review by fishjojo — APPROVED** [review]

**Pull request merged**